Baseline-comparison API

fiedl opened the pull request:
In the context of the [baseline-comparison feature](https://community.openproject.org/projects/openproject/work_packages/26448), this pull request extends the work-packages api to support a `timestamps` parameter and to return the `attributesByTimestamp` for the different timestamps embedded within the api response.

## Context

![Component_Overview-7](https://user-images.githubusercontent.com/1679688/220207617-a44eec78-d8af-4c2c-a393-ba4ede80b780.png)

## Usecase 1: Compare work packages to a baseline state through the api

To retrieve a collection of work packages, use the `/api/v3/work_packages` endpoint, e.g. https://community.openproject.org/api/v3/work_packages.json.

In order to query different points in time, provide a `timestamps` parameter with comma-separated [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) strings.

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "total": 1,
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "id": 1528,
        "subject": "Current subject of the work package",
        // other attributes ...,
        "_links": {
          "self": {
            "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z,PT0S"
          }
        },
        "_meta": {
          "matchesFilters": true,
          "exists": true,
          "timestamp": "PT0S"
        },
        "_embedded": {
          "attributesByTimestamp": [
            {
              "subject": "Original subject of the work package",
              "_meta": {
                "matchesFilters": true,
                "exists": true,
                "timestamp": "2022-01-01T00:00:00Z"
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z"
                }
              },
            },
            {
              "_meta": {
                "matchesFilters": true,
                "exists": true,
                "timestamp": "PT0S"
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=PT0S"
                }
              }
            }
          ],
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "/api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S"
    }
  }
}
```

In this example ruby script, the work packages are retrieved at a baseline date and in their current state. Then the subject of the first work package is compared with respect to the baseline date and the current state.

```ruby
# Define timestamps
baseline_timestamp = "2022-01-01T00:00:00Z"
current_timestamp = "PT0S"
timestamps = [baseline_timestamp, current_timestamp]

# Retrieve work packages
url = "https://community.openproject.org/api/v3/work_packages?timestamps=#{timestamps.join(',')}"
response = JSON.parse(RestClient.get(url), object_class: OpenStruct)
work_packages = response.dig("_embedded", "elements")

# Extract differing baseline attributes
work_package = work_packages.first
baseline_attributes = work_package.dig("_embedded", "attributesByTimestamp").first

# Compare baseline state to current state of the work package
if baseline_attributes.subject.present? and baseline_attributes.subject != work_package.subject
  puts "The subject of the work package has changed."
  puts "Subject at the baseline time: #{baseline_attributes.subject}"
  puts "Current subject:              #{work_package.subject}"
end

# Check existence
puts "The work package did exist at the baseline time." if baseline_attributes.dig("_meta", "exists")
puts "The work package exists at the current time." if work_package.dig("_meta", "exists")

# Check filter matching
puts "The work package matches the query filters at the baseline time." if baseline_attributes.dig("_meta", "matchesFilters")
puts "The work package matches the query filters at the current time." if work_package.dig("_meta", "matchesFilters")
```

## Usecase 2: Compare work packages to a baseline state through the user interface

<img width="959" alt="Screenshot_2022-05-02_at_15 33 37" src="https://user-images.githubusercontent.com/1679688/206807823-6bb25bbc-8d36-4c2e-adbf-ce1ff3377f19.png">

```haml
- baseline_timestamp = "2022-01-01T00:00:00Z"

- url = "https://community.openproject.org/api/v3/work_packages?baseline=#{baseline_timestamp}"
- response = JSON.parse(RestClient.get(url), object_class: OpenStruct)
- work_packages = response.dig("_embedded", "elements")

%table.work_packages
  - for work_package in work_packages do
    - baseline_attributes = work_package.dig("_embedded", "attributesByTimestamp").first
  
    - matches_filters_now = work_package.dig("_meta", "matchesFilters")
    - matches_filters_at_baseline = baseline_attributes.dig("_meta", "matchesFilters")
    
    - css_class = "green" if matches_filters_now and not matches_filters_at_baseline
    - css_class = "red" if matches_filters_at_baseline and not matches_filters_now

    %tr.work_package{class: css_class}
      %td.id= work_package.id
      %td.subject
        - if baseline_attributes.try(:subject).present? and work_package.subject.present? and (baseline_attributes.subject != work_package.subject)
          %div.baseline_value= baseline_attributes.subject
        - if work_package.subject.present?
          %div.value= work_package.subject
        - elsif baseline_attributes.try(:subject).present?
          %div.deleted_value= baseline_attributes.subject
```

This style scheme works without checking the existence of the work packages. If required, however, the existence can be read out from the meta section:

```haml
- exists_now = work_package.dig("_meta", "exists")
- exists_at_baseline = baseline_attributes.dig("_meta", "exists")
```

## Usecase 3: Access historic attributes through a wrapper object

This pull request provides the `Journable::WithHistoricAttributes` wrapper to decorate the work packages with their historic attributes (`attributes_by_timestamp`) as well as with meta information on whether the work packages match the query filters (`matches_query_filters_at_timestamps`). ➝ https://github.com/opf/openproject/pull/11783/commits/6dcfea8cc82cb52a82bc3b5356403e7e8cd6daf5

```ruby
# Wrap single work package
timestamps = [Timestamp.parse("2022-01-01T00:00:00Z"), Timestamp.parse("PT0S")]
work_package = WorkPackage.find(1)
work_package = Journable::WithHistoricAttributes.wrap_one(work_package, timestamps:)

# Wrap multiple work packages
timestamps = query.timestamps
work_packages = query.results.work_packages
work_packages = Journable::WithHistoricAttributes.wrap_multiple(work_packages, timestamps:)

# Access historic attributes at timestamps after wrapping
work_package = Journable::WithHistoricAttributes.wrap(work_package, timestamps:)
work_package.subject  # => "Subject at PT0S (current time)"
work_package.attributes_by_timestamp["2022-01-01T00:00:00Z"].subject  # => "Subject at 2022-01-01 (baseline time)"

# Check at which timestamps the work package matches query filters after wrapping
query.timestamps  # => [<Timestamp 2022-01-01T00:00:00Z>, <Timestamp PT0S>]
work_package = Journable::WithHistoricAttributes.wrap(work_package, query:)
work_package.matches_query_filters_at_timestamps  # => [<Timestamp 2022-01-01T00:00:00Z>]

# Include only changed attributes in payload
# i.e. only historic attributes that differ from the work_package's attributes
timestamps = [Timestamp.parse("2022-01-01T00:00:00Z"), Timestamp.parse("PT0S")]
work_package = Journable::WithHistoricAttributes.wrap(work_package, timestamps:, include_only_changed_attributes: true)
work_package.attributes_by_timestamp["2022-01-01T00:00:00Z"].subject  # => "Subject at 2022-01-01 (baseline time)"
work_package.attributes_by_timestamp["PT0S"].subject  # => nil

# Simplified interface for two timestamps
query.timestamps  # => [<Timestamp 2022-01-01T00:00:00Z>, <Timestamp PT0S>]
work_package = Journable::WithHistoricAttributes.wrap(work_package, query:)
work_package.baseline_timestamp  # => [<Timestamp 2022-01-01T00:00:00Z>]
work_package.current_timestamp  # => [<Timestamp PT0S>]
work_package.matches_query_filters_at_baseline_timestamp?
work_package.matches_query_filters_at_current_timestamp?
work_package.baseline_attributes.subject # => "Subject at 2022-01-01 (baseline time)"
work_package.subject  # => "Subject at PT0S (current time)"
```

## Work to be done

- [x] Implement wrapper `Journable::WithAttributesAtTimestamps` to decorate work packages with historic attributes and meta information. ➝ https://github.com/opf/openproject/pull/11783/commits/6dcfea8cc82cb52a82bc3b5356403e7e8cd6daf5
  - [x] Better name: `Journable::WithHistoricAttributes` ➝ https://github.com/opf/openproject/pull/11783/commits/18339f3463dc0cfa0fa95aff9b0a4aa60c77a772

### `GET /api/v3/work_packages` endpoint

- [x] Parse comma-separated `timestamps` parameter with [ISO_8601](https://en.wikipedia.org/wiki/ISO_8601) ➝ https://github.com/opf/openproject/pull/11783/commits/151adf555135b0e53a6a62df8b170b97079b2495
- [x] Support special strings like `"now"` or `"-1y"` ➝ https://github.com/opf/openproject/pull/11783/commits/3231f1d7f990453abee7804bc8f1ce76c5230410
- [x] Pass parameters to the `Query` object ➝ https://github.com/opf/openproject/pull/11783/commits/ffca847e65c8e1947818a1a461879d29577c97a9
- [x] Update `./docs/api/apiv3/paths/work_package.yml` ➝ https://github.com/opf/openproject/pull/11783/commits/f4585baa3f71dfd65e45652e01469f6182bffb38
- [x] Include `attributesByTimestamp` in the api response ➝ https://github.com/opf/openproject/pull/11783/commits/edb3cd0381db9d7f4f5477500fb9b9c140be564f
- [x] Add `_links` with timestamps
- [x] Add `_meta` information ➝ https://github.com/opf/openproject/pull/11783/commits/4e8a84ffbd93bae6c28b632ed885309d7667fed6
- [x] Spec: provide `timestamps` parameter and check `baselineAttributes` in response ➝ https://github.com/opf/openproject/pull/11783/commits/d5bea80141838babb1b84f7d41ef3180cb53cd3f

### `GET /api/v3/work_packages/:id` endpoint

- [x] Parse comma-separated `timestamps` parameter ➝ https://github.com/opf/openproject/pull/11783/commits/8dce4df30ef87e7d46932514d54b358a5aa5a084
- [x] Update `./docs/api/apiv3/paths/work_package.yml` ➝ https://github.com/opf/openproject/pull/11783/commits/f4585baa3f71dfd65e45652e01469f6182bffb38
- [x] Include `attributesByTimestamp` in the api response ➝ https://github.com/opf/openproject/pull/11783/commits/edb3cd0381db9d7f4f5477500fb9b9c140be564f
- [x] Add `_links` with timestamps
- [x] Add `_meta` information ➝ https://github.com/opf/openproject/pull/11783/commits/4e8a84ffbd93bae6c28b632ed885309d7667fed6
- [x] Spec: provide `timestamps` parameter and check `attributesByTimestamp` in response ➝ https://github.com/opf/openproject/pull/11783/commits/8dce4df30ef87e7d46932514d54b358a5aa5a084

## Next steps

- Use representer to render the `attributesByTimestamp`: https://github.com/opf/openproject/pull/12107
- Improve performance

## API Documentation

- https://www.openproject.com/docs/api/introduction
- https://www.openproject.org/docs/api/endpoints/work-packages/
- http://localhost:3000/api/docs 
- [./docs/api/apiv3/openapi-spec.yml](https://github.com/opf/openproject/tree/dev/docs/api/apiv3/openapi-spec.yml)

## See also

- **Implementation Overview**: https://github.com/fiedl/openproject/issues/4
- [**Epic #26448**: Track changes in the work package list and Gantt view (baseline comparison)](https://community.openproject.org/projects/openproject/work_packages/26448)

fiedl committed:
baseline api: parse timestamps in ParseQueryParamsService

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: pass timestamps to query

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: handle special-value timestamps like "now" or "2y"

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: provide wrapper `Journable::WithAttributesAtTimestamps`

to decorate work packages with historic attributes and meta information.

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rubocop refactoring

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rename wrapper to `Journable::WithHistoricAttributes`

https://github.com/opf/openproject/pull/11783

fiedl commented:
## Nomenclature: How to call the historic-attributes hash? ➝ `attributesByTimestamp`

The historic attributes are embedded in the work-package results. They are collected in a hash where the key is timestamp string.

```ruby
{
  "_type": "WorkPackage",
  "_embedded": {
    "attributesByTimestamp": {
      "PT-1Y": {
        "subject": "Bar"
      }
    }
  }
  "subject": "Foo"
}
```

How should the hash be called?

- `historicAttributes`: As the timestamp may also be the present time, I'm dropping the "historic" from the name.
- `attributes`: The object itself is no collection of attributes as "attributes" would suggest, but a hash where the key is the timestamp string.
- `attributesAtTimestamps`: The plural-s would suggest that the keys could be several timestamps, but actually the key is a single timestamp string.
- `attributesAtTimestamp`: This would raise the question "which timestamp?"
- `attributesByTimestamp`: This is the clearest name:
  - The object is a hash where each key is a timestamp
  - The values are attributes, i.e. a hash of attributes

"attributes by foo": Is foo singular or plural? ➝ singular

https://www.oxfordlearnersdictionaries.com/definition/english/sort_2

> sort *something* by *something*: Events are sorted by date and category.

=> **`attributesByTimestamp`**

fiedl commented:
## Redundant `baselineAttributes`

In addition to the `attributesByTimestamp`, the work package's `_embedded` section holds a `baselineAttributes` hash, which provides access to the attributes at the baseline timestamp without knowing the baseline timestamp. This is a redundant convenience accessor and can be dropped if needed.

```ruby
{
  "_type": "WorkPackage",
  "_embedded": {
    "baselineAttributes": {
      "subject": "Bar",
      "_meta": {
        "timestamp": "P-1Y"
      }
    },
    "attributesByTimestamp": {
      "P-1Y": {
        "subject": "Bar"
      }
    }
  }
  "subject": "Foo"
}
```

Because the hash key does not have the information of the timestamp, I'm including the timestamp string as `_meta` information. In principle, the api consumer does know the timestamp string because she has provided it as query parameter. This is just a verification method and can be dropped if needed.

fiedl committed:
Journable::Timestamps: columns missing in the journal-data table are interpreted as null

https://github.com/opf/openproject/pull/11243

fiedl committed:
baseline api: rename accessor to `Journable::WithHistoricAttributes#attributes_by_timestamp`

https://github.com/opf/openproject/pull/11783#issuecomment-1356777648

fiedl committed:
baseline api: pass `query` and `timestamps` to the `WorkPackageCollectionRepresenter`

https://github.com/opf/openproject/pull/11783

fiedl commented:
## How to pass `timestamps` and `query` to the `Journable::WithHistoricAttributes`?

![Bildschirm­foto 2022-12-18 um 20 35 46](https://user-images.githubusercontent.com/1679688/208315840-80295459-f110-4f24-9de3-1864c5bae36d.png)

fiedl committed:
baseline api: add `Journable::WithHistoricTimestamps` attributes through eager-loading wrapper

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: redefine `inspect` on wrapper objects to help debugging

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: fixing accessor name in specs

https://github.com/opf/openproject/pull/11783#issuecomment-1356777648

fiedl committed:
baseline api: add `attributesByTimestamp` and `baselineAttributes` to work package representer

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: have the representer include the baseine-attributes properties only when they are present

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: the work-package-collection self link includes the absolute timestamps

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: add `timestamps` only to `query` hash if present

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: add self link to embedded baseline attributes

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: don't include timestamps in links when only `Timestamp.now` is queried

because this is the default case when no timestamp is given

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: don't mock the eager-loading wrapper

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: include as meta information at which timestamps the work package matches the query filters

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: pass timestamps to representer on show endpoint

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: include `matchesFilters` meta info only if a query is given

Otherwise, there are not filters to match and this field is meaningless.

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: make `Journable::WithHistoricAttributes.wrap` detect single object and multiple objects

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rename `matches_query_at_timestamps` to `matches_query_filters_at_timestamps`

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: pass timestamps separately to representer rather than with the query params

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rubocop refactoring

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: use `Journable::WithHistoricAttributes include_only_changed_attributes: true` for api requests

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: fix naming of `matches_query_filters_at_timestamp`

which should be plural because there can be several filters in the query.

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: include `_meta/matchesFilters` in the base work package

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rubocop refactoring

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: adding request specs for `/work_packages?timestamps=...`

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rubocop refactoring

fiedl commented:
## Some more things to do

- [x] [Extend the open-api specs](https://github.com/opf/openproject/pull/11783#issuecomment-1437153632)
- [x] Request spec with a work package that has not changed at all ➝ https://github.com/opf/openproject/pull/11783/commits/1219ac3c0ff4492fc9d12a5d22327b8eda8860d2
- [x] Request spec with a work package that has not been present at the baseline date ➝ https://github.com/fiedl/openproject/commit/f7b57341f3bd895cc536e0c266e038f47af903a9
- [x] Test it in the browser
  - [x] Fix caching issue ➝ https://github.com/opf/openproject/pull/11783/commits/acb05262d1d27a9b20fcd5351ca7964178ab8ac3
- [x] Bring https://github.com/fiedl/openproject/issues/4 up to date and collect things left to do later
- [x] Remove `baselineAttributes` in favour of `attributesByTimestamp` ➝ https://github.com/opf/openproject/pull/11783/commits/22c95c45c82f5eb10de961643d3b96adb7581d23
- [x] Add spec to ensure backwards compatibility: If no `timestamps` are given, none of the new json objects should be present ➝ https://github.com/opf/openproject/pull/11783/commits/18dbf27139bb8f77cdc61c4f16263f183275e28a

fiedl commented:
## Check for existence at the baseline time

Consider a work package that is created after the baseline time.

### (a) Info from the meta section

```ruby
baseline_timestamp = "2022-01-01T00:00:00Z"

url = "https://community.openproject.org/api/v3/work_packages?baseline=#{baseline_timestamp}"
response = JSON.parse(RestClient.get(url), object_class: OpenStruct)
work_packages = response.dig("_embedded", "elements")

work_package = work_packages.first
baseline_attributes = work_package.dig("_embedded", "baselineAttributes")

if baseline_attributes.dig("_meta", "exists")
  puts "The work package already existed at the baseline time"
else
  puts "The work package did not exist at the baseline time"
end
```

### (b) Implicit info from the attributes being present

```ruby
baseline_timestamp = "2022-01-01T00:00:00Z"

url = "https://community.openproject.org/api/v3/work_packages?baseline=#{baseline_timestamp}"
response = JSON.parse(RestClient.get(url), object_class: OpenStruct)
work_packages = response.dig("_embedded", "elements")

work_package = work_packages.first
baseline_attributes = work_package.dig("_embedded", "baselineAttributes")

if baseline_attributes.present?
  puts "The work package already existed at the baseline time"
else
  puts "The work package did not exist at the baseline time"
end
```

- This approach does not work in the main section because the main section always exists. One would need to check the embedded section, i.e. `attributesAtTimestamps`

### (c) Using a `404` status code

- This would hide baseline attributes when the work packages has existed, but does not exist anymore

fiedl committed:
baseline api: track existence as meta attribute

https://github.com/opf/openproject/pull/11783#issuecomment-1374897874

fiedl commented:
## `baselineAttributes` vs. `attributesByTimestamp`

Having only the attributes and meta information of the baseline time in the `baselineAttributes` embedded object is specific to the problem where we have the baseline time and the current time as `timestamp`.

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "total": 1,
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "id": 1528,
        "subject": "Current subject of the work package",
        "_links": {
          "self": {
            "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z,PT0S"
          }
        },
        "_meta": {
          "matchesFilters": true
        },
        "_embedded": {
          "baselineAttributes": {
            "subject": "Original subject of the work package",
            "_meta": {
              "matchesFilters": true,
              "timestamp": "2022-01-01T00:00:00Z"
            },
            "_links": {
              "self": {
                "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z"
              }
            },
          }
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "/api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S"
    }
  }
}
```

The more general approach has the attributes and meta information for all timestamps in the `attributesByTimestamp` hash. However, this makes getting just the baseline attributes a little more cumbersome because one needs to provide the baseline time as key.

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "total": 1,
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "id": 1528,
        "subject": "Current subject of the work package",
        "_links": {
          "self": {
            "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z,PT0S"
          }
        },
        "_meta": {
          "matchesFilters": true
        },
        "_embedded": {
          "attributesByTimestamp": {
            "2022-01-01T00:00:00Z": {
              "subject": "Original subject of the work package",
              "_meta": {
                "matchesFilters": true
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z"
                }
              },
            },
            "PT0S": {
              "_meta": {
                "matchesFilters": true
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=PT0S"
                }
              }
            }
          },
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "/api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S"
    }
  }
}
```

A compromise would be to make the `attributesByTimestamp` an array rather than a hash. The order of the array is the order of `timestamps`. As a cross check, the `timestamp` is given within the `_meta` section. This way, one can get the baseline attributes just by accessing the first element of the `attributesByTimestamp` array.

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "total": 1,
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "id": 1528,
        "subject": "Current subject of the work package",
        "_links": {
          "self": {
            "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z,PT0S"
          }
        },
        "_meta": {
          "matchesFilters": true
        },
        "_embedded": {
          "attributesByTimestamp": [
            {
              "subject": "Original subject of the work package",
              "_meta": {
                "matchesFilters": true,
                "timestamp": "2022-01-01T00:00:00Z"
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z"
                }
              },
            },
            {
              "_meta": {
                "matchesFilters": true,
                "timestamp": "PT0S"
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=PT0S"
                }
              }
            }
          ],
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "/api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S"
    }
  }
}
```

fiedl committed:
baseline api: make `attributesByTimestamp` an array rather than a hash

https://github.com/opf/openproject/pull/11783#issuecomment-1379678721

ulferts commented on app/models/timestamp.rb:
```diff
@@ -0,0 +1,169 @@
+    years = string.scan(/(\d+)y/).flatten.first
+    months = string.scan(/(\d+)m/).flatten.first
+    weeks = string.scan(/(\d+)w/).flatten.first
+    days = string.scan(/(\d+)d/).flatten.first
```
`string.scan(/(\d+)[delimiter]/).flatten.first` could be extracted into a method.

ulferts replied:
I don't care if the the custom time-references like `P1Y` stay in here as well. Might get removed later on but I see no harm in having them.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/debc37df078b07e625502700af6750c7bcf7c096

fiedl committed:
Merge 'dev' into https://github.com/opf/openproject/pull/11783

fiedl committed:
Merge 'dev' into https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: adding spec with a work package that has not changed at all

https://github.com/opf/openproject/pull/11783#issuecomment-1360596592

fiedl commented:
## OpenAPI Specs

*2022-06-23*

- ./docs/api/apiv3/openapi-spec.yml
- https://github.com/opf/openproject/blob/dev/docs/api/apiv3/openapi-spec.yml
- [OpenAPI 3.0 Specification](https://swagger.io/specification/)
- https://www.openproject.com/docs/api/introduction
- http://localhost:3000/api/docs 

I'm drafting the documentation as follows:

fiedl commented:
# Baseline Comparisons

Baseline comparisons allow to compare work packages or collections of work packages with respect to different points in time.

This helps to answer questions like:

- Which work packages match a certain set of filters today, which work packages match this set of filters at a certain earlier point in time?
- Which properties of these work packages have changed with respect to these points in time?

This tool can be used to analyze how a project plan has changed with respect to a certain baseline date.

## Requesting Work Packages for Different Timestamps

The work-packages API supports a `timestamps` parameter to gather information about a single work package or a collection of work packages for several points in time.

```
GET /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
```

```
GET /api/v3/work_packages/123?timestamps=2022-01-01T00:00:00Z,PT0S
```

Each timestamp should be given as an [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) string, either an absolute date and time with timezone, e.g. `"2022-01-01T00:00:00Z"`, or a relative timestamp utilizing the [ISO-8601-Duration](https://en.wikipedia.org/wiki/ISO_8601#Durations) format, e.g. `"P-1Y"`, which is composed of an initial `"P"` for "Period", and a duration. `"P-1Y"` is interpreted as the relative timestamp "1 year ago".

Several timestamps should be passed as comma-separated list of these ISO-8601 strings to the `timestamps` parameter, e.g. `"2022-01-01T00:00:00Z,PT0S"`.

The timestamps should be given in ascending temporal order, such that the first timestamp is the **baseline** timestamp, and the last timestamp is the **current** timestamp.

## Response Overview

When providing a `timestamps` parameter, the response has several additional properties:

| Property                | Description                                                                            | Type             | Further information                                 |
| ----------------------- | -------------------------------------------------------------------------------------- | ---------------- | --------------------------------------------------- |
| `timestamp`             | The requested timestamp corresponding to the surrounding embedded object               | String           | Section *[Timestamps](#timestamps)* below           |
| `attributesByTimestamp` | Attributes and meta information of the work packages at the respective timestamps      | Array of Objects | Section *[Attributes](#attributes)* below           |
| `matchesFilters`        | Marks whether the work package matches the filter criteria at the respective timestamp | Bool             | Section *[Filter Matching](#filter-matching)* below |
| `exists`                | Marks whether the work package exists at the respective timestamp                      | Bool | Section *[Existence](#existence)* below             |

Each work-package element has the `attributesByTimestamp` as `_embedded` section.

The properties `timestamp`, `matchesFilters`, and `exists` are wrapped in a `_meta` section, which is added to each work-package element as well as to each element of the `attributesByTimestamp` array.

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "total": 1,
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "id": 1528,
        "subject": "Current subject of the work package",
        // other attributes ...,
        "_links": {
          "self": {
            "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z,2023-03-01T01:37:10Z"
          }
        },
        "_meta": {
          "matchesFilters": true,
          "exists": true,
          "timestamp": "PT0S"
        },
        "_embedded": {
          "attributesByTimestamp": [
            {
              "subject": "Original subject of the work package",
              "_meta": {
                "matchesFilters": true,
                "exists": true,
                "timestamp": "2022-01-01T00:00:00Z"
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=2022-01-01T00:00:00Z"
                }
              },
            },
            {
              "_meta": {
                "matchesFilters": true,
                "exists": true,
                "timestamp": "PT0S"
              },
              "_links": {
                "self": {
                  "href": "/api/v3/work_packages/1528?timestamps=2023-03-01T01:37:10Z"
                }
              }
            }
          ],
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "/api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,2023-03-01T01:37:10Z"
    }
  }
}
```

## Meta Information

Each `_meta` section describes the surrounding object of the meta section, which may be the main work-package object, or an attributes object within the `attributesByTimestamp` array.

Note that the `_meta` information of the most current (rightmost) timestamp is redundant: It is given as `_meta` section of the main work-package object as well as `_meta` section of the last object within the `attributesByTimestamp` array.

## Timestamps

Each `_meta` section contains a `timestamp` property, which is to the requested `timestamp` corresponding to the object the `_meta` section describes.

The `timestamp` has the same ISO-8601 format as in the `timestamps` request parameter and preserves the absolute or relative character of the requested timestamp.

Furthermore, each self link corresponding to an earlier point in time has also a `timestamps` request parameter added to it, which is converted to an absolute ISO-8601 string at the execution time of the query. (`"2023-03-01T01:37:10Z` in the above example, which has been executed at that time.)

## Attributes

To read out the attributes of the work packages at the current timestamp (the last of the given `timestamps`), check the attributes of the work-package objects. To read out the attributes of the work packages at the other given timestamps, check the attributes within `"_embedded"` section `"attributesAtTimestamp"`.

To save bandwidth, only attributes that differ from the ones in the main work-package object are included in the `"attributesByTimestamp"`. Attributes with the same value as in the main work-package object are not included in the `"attributesByTimestamp"` section.

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "subject": "Current subject of the work package",
        "_meta": {
          "timestamp": "PT0S"
        },
        "_embedded": {
          "attributesByTimestamp": [
            {
              "subject": "Original subject of the work package",
              "_meta": {
                "timestamp": "2022-01-01T00:00:00Z"
              }
            },
            {
              "_meta": {
                "timestamp": "PT0S"
              }
            }
          ],
        }
      }
    ]
  }
}
```

In the above example, the last of the given `timestamps` is `"PT0S"` (which means 0 seconds ago, i.e. now). The work-package attributes at this time are included in the main work-package object. The `"subject"` of the work package at the timestamp `"PT0S"` (now) is `"Current subject of the work package"`.

The `"_embedded"` section `"attributesByTimestamp"` has an array entry for the timestamp `"PT0S"`, which is the last array entry. Because the value of the `"subject"` is the same as up in the main work-package object, the `"subject"` attribute is left out in the `"attributesByTimestamp"` for the timestamp `"PT0S"`. The `"subject"` of the work package at the timestamp `"PT0S"` (now) is `"Current subject of the work package"`.

The `"_embedded"` section `"attributesByTimestamp"` has an array entry for the baseline timestamp `"2022-01-01T00:00:00Z"`, which is the first array entry. The `"subject"` of the work package at the timestamp `"2022-01-01T00:00:00Z"` is `"Original subject of the work package"`. It is included in the  `"attributesByTimestamp"` for the timestamp `"2022-01-01T00:00:00Z"` because it differs from the `"subject"` in the main work-package object, which is `"Current subject of the work package"`.

## Filter Matching

The work-packages API supports filtering the query results by one or several search criteria. See: [Filters](../filters)

To find out whether a work package matches the given set of filter criteria at a certain timestamp, check the `"matchesFilters"` property in the `"_meta"` section for that timestamp:

```json
// /api/v3/work_packages?filters=...&timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "_meta": {
          "matchesFilters": true,
          "timestamp": "PT0S"
        },
        "_embedded": {
          "attributesByTimestamp": [
            {
              "_meta": {
                "matchesFilters": false,
                "timestamp": "2022-01-01T00:00:00Z"
              }
            },
            {
              "_meta": {
                "matchesFilters": true,
                "timestamp": "PT0S"
              }
            }
          ],
        }
      }
    ]
  }
}
```

In the above example, the work package matches the filter criteria at the timestamp `"PT0S"`, but does not match the filter criteria at the timestamp `"2022-01-01T00:00:00Z"`.

In another example, it might be the other way around: The work package could match the filter criteria (`"matchesFilters": true`) at the baseline timestamp, but not match the filter criteria anymore (`"matchesFilters": false`) at the current timestamp.

The work package is included in the returned collection if it matches the filter criteria at least at one of the requested timestamps.

## Existence

To find out whether a work package has existed at a requested timestamp, check the `"exists"` property in the `"_meta"` section for that timestamp:

```json
// /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
{
  "_type": "WorkPackageCollection",
  "_embedded": {
    "elements": [
      {
        "_type": "WorkPackage",
        "_meta": {
          "exists": true,
          "timestamp": "PT0S"
        },
        "_embedded": {
          "attributesByTimestamp": [
            {
              "_meta": {
                "exists": false,
                "timestamp": "2022-01-01T00:00:00Z"
              }
            },
            {
              "_meta": {
                "exists": true,
                "timestamp": "PT0S"
              }
            }
          ],
        }
      }
    ]
  }
}
```

In the above example, the work package exists at the timestamp `"PT0S"`, but has not existed at the timestamp `"2022-01-01T00:00:00Z"`.

In another example, it might be the other way around: The work package could exist (`"exists": true`) at the baseline time, but could have been deleted after that time such that it does not exist (`"exists": false`) at the current time. Please note, however, that OpenProject does not support [soft deletion](https://community.openproject.org/projects/openproject/work_packages/40015), yet. Currently, when a work package is deleted, its history is deleted as well, so that its history cannot be retrieved through the baseline API anymore.

The work package is included in the returned collection if it has existed at least at one of the requested timestamps.

## Usage Example

In this example ruby script, the work packages are retrieved at a baseline date and in their current state. Then the subject of the first work package is compared with respect to the baseline date and the current state.

```ruby
# Define timestamps
baseline_timestamp = "2022-01-01T00:00:00Z"
current_timestamp = "PT0S"
timestamps = [baseline_timestamp, current_timestamp]

# Retrieve work packages
url = "https://community.openproject.org/api/v3/work_packages?timestamps=#{timestamps.join(',')}"
response = JSON.parse(RestClient.get(url), object_class: OpenStruct)
work_packages = response.dig("_embedded", "elements")

# Extract differing baseline attributes
work_package = work_packages.first
baseline_attributes = work_package.dig("_embedded", "attributesByTimestamp").first

# Compare baseline state to current state of the work package
if baseline_attributes.subject.present? and baseline_attributes.subject != work_package.subject
  puts "The subject of the work package has changed."
  puts "Subject at the baseline time: #{baseline_attributes.subject}"
  puts "Current subject:              #{work_package.subject}"
end

# Check existence
puts "The work package did exist at the baseline time." if baseline_attributes.dig("_meta", "exists")
puts "The work package exists at the current time." if work_package.dig("_meta", "exists")

# Check filter matching
puts "The work package matches the query filters at the baseline time." if baseline_attributes.dig("_meta", "matchesFilters")
puts "The work package matches the query filters at the current time." if work_package.dig("_meta", "matchesFilters")
```

fiedl committed:
baseline api: adding api documentation

https://github.com/opf/openproject/pull/11783#issuecomment-1437582041

fiedl committed:
Merge 'dev' into https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: add timestamps to cache key

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: show `attributesByTimestamp` and `_meta` when using `timestamps`

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: fixing `exists` by avoiding subquery for now

Otherwise, `exists` shows always `false` currently in the browser.

https://github.com/opf/openproject/pull/11783

This requires permanent fixing: https://github.com/fiedl/openproject/issues/4

fiedl committed:
baseline api: removing redundant `baselineAttributes`

in favour of `attributesByTimestamp`

discussion: https://github.com/opf/openproject/pull/11783#issuecomment-1379678721

fiedl committed:
baseline api: adding spec for backwards compatibility

The extra attributes should not be included in the response if the timestamps feature is not used.

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: fixing open-api spec:

The `timestamps` param is a query param, not included within the path.

https://github.com/opf/openproject/actions/runs/4228018945/jobs/7343037088

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: show "_meta" "matchesFilters" only when a query is passed

https://github.com/opf/openproject/actions/runs/4228018945/jobs/7343036962

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rubocop refactoring

https://github.com/opf/openproject/actions/runs/4228018947/jobs/7343036712

https://github.com/opf/openproject/pull/11783

ulferts commented on app/models/journable/with_historic_attributes.rb:
```diff
@@ -0,0 +1,225 @@
+    __getobj__.inspect.gsub(/#<(.+)>/m, "#<#{self.class.name} \\1>")
+  end
+
+  class NotImplemented < StandardError; end
```
I'll skip reviewing it as I reworked it extensively in #12107. But the interface is unchanged there, only some methods added that I needed in the WorkPackageRepresenter and for the eager loading. Internally it changed to a structure where a `Loader` (Registry) object is used so that the information is fetched efficiently from the database.

ulferts commented on app/services/api/v3/parse_query_params_service.rb:
```diff
@@ -37,8 +37,9 @@ def call(params)
         return json_parsed unless json_parsed.success?
 
         parsed = parsed_params(params)
+        return parsed unless parsed.success?
```
It reads easier with `if parsed.failure?`. At least to me.

fiedl replied:
👍 I had copied line 37 above. I'm changing both occurrences, now. ➝ https://github.com/opf/openproject/pull/11783/commits/0861cb95ebca67745522bc7ac4d8962dc3c57a67

ulferts commented on lib/api/v3/work_packages/eager_loading/historic_attributes.rb:
```diff
@@ -0,0 +1,87 @@
+    def matches_query_filters_at_timestamp?(timestamp)
+      matches_query_filters_at_timestamps.any? ? matches_query_filters_at_timestamps.include?(timestamp) : nil
+    end
+  end
```
This I also reworked extensively in #11783. The main driver for this is the usage of a subclass of the `WorkPackageRepresenter` to render the `attributesAtTimestamp`. Because of that, a WorkPackage-compatible object needed to be handled on parsing the `attributesAtTimestamp` collection.

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -317,6 +330,33 @@ def initialize(model, current_user:, embed_links: false)
+                 if: ->(*) {
+                   respond_to? :matches_query_filters_at_timestamps \
+                   and respond_to? :timestamps \
+                   and timestamps != [Timestamp.now]
```
In #11783 I extracted this check into a method checking `timestamps.any?(&:historic?)`. I believe this check to suffice and be quite precise.

fiedl replied:
Actually, I had tried something similar. But I always got errors like "*undefined method `timestamps_active?' for #<WorkPackage*". I guess, I did not fully grasp the binding of the callbacks. I'm leaving this for now in this pull request, but I'm happy with you extracting the check into a method. 👍

ulferts commented on lib/api/v3/work_packages/work_package_collection_representer.rb:
```diff
@@ -39,10 +41,25 @@ def initialize(models,
+          # the query parameters passed to the API.
+          #
+          # TODO: Should we rename `_query` to `query_object`?
+          # TODO: Should we rename `query` to `query_params`?
```
I'd rename what is currently `query` to `query_params`. It is in line with Rails (`params`)  and most devs will understand it. 'query', within the scope of OpenProject, I'd use for an instance of the `Query` class.

fiedl replied:
I've renamed these parameters in https://github.com/opf/openproject/pull/11783/commits/a24a06fefa08c612d698d9a3ba9cedfb777cad7f as suggested.

This also affects `API::Decorators::OffsetPaginatedCollection`. All specs are green, but let's watch out for unintended side effects.

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -642,14 +713,17 @@ def json_cache_key
            I18n.locale,
            json_key_representer_parts,
            represented.cache_checksum,
+           timestamps,
```
Good addition. I did not find any tests for it in the specs. Please add some.

fiedl replied:
I have added specs for caching. ➝ https://github.com/opf/openproject/pull/11783/commits/a18925e229130746e3c940fe88386fe6b6924ae8

Actually, adding the `timestamps` to the `json_cache_key` is not necessary currently, because the relevant json properties are already marked with `uncacheable: true`.

- I'm removing the `timestamps` from the `json_cache_key` for now. ➝ https://github.com/opf/openproject/pull/11783/commits/b35df5b1fde95c4c7a1c832d4a366fec0af57c73
- I've added the task of re-enabling caching in the scenarios we've [discussed today](https://matrix.to/#/!JNelJmTpjLHzTwgpaV:fiedl.net/$frC4BP9tLst2M-Fhk6qeyb10NZcfYKuEMH16o_k1G0E?via=fiedl.net&via=openproject.org) to the [follow-up steps](https://github.com/fiedl/openproject/issues/4).

ulferts commented on lib/api/v3/work_packages/work_package_collection_representer.rb:
```diff
@@ -39,10 +41,25 @@ def initialize(models,
-                       embed_schemas: false)
+                       embed_schemas: false,
+                       timestamps: [],
+                       _query: nil)
```
Admittedly, I don't like the interface becoming this large. But also admittedly, I don't have an idea to avoid this because of the various cases that need to be handled (based on query, based on just work packages array or AR::Relation). So this is what it is. Just writing this to maybe trigger a better idea by somebody.

fiedl replied:
When implementing this, I've thought about maybe leaving out the `timestamps` parameter here, because it is also available through the `query`.  But I liked the explicit mention of the `timestamps`. Also, it made the specs slightly nicer.

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -438,6 +478,37 @@ def initialize(model, current_user:, embed_links: false)
+                                             'exists': exists_at_timestamps.include?(timestamp)
+                                           }.compact
+                                         })
+                     attrs
```
Rewritten in #11783. A subclass of this Representer is used that will render only the properties changed between the root work package and the state at the various timestamps. That way all the property rendering logic of the `WorkPackageRepresenter` can be reused.

ulferts commented on spec/requests/api/v3/work_packages/show_resource_spec.rb:
```diff
@@ -206,4 +206,94 @@
+        end
+
+        before do
+          WorkPackage.destroy_all
```
Those line should not be necessary (in none of the test cases) since the test DB should not contain any WorkPackages of previous tests.

fiedl replied:
Thanks

ulferts commented on lib/api/v3/work_packages/work_package_eager_loading_wrapper.rb:
```diff
@@ -35,23 +35,25 @@ def wrapped?
-          def wrap_and_apply(work_packages, container_classes)
+          def wrap_and_apply(work_packages, container_classes, timestamps:, query:)
             containers = container_classes
                          .map { |klass| klass.new(work_packages) }
```
I'd rather extend the initializer interface of all `container_classes` to take additional options. In most cases, they will simply be ignored. This is way more extensible and avoids the cumbersome `if container.respond_to?(...)` later on.

fiedl replied:
I've extended the initializer of `API::V3::WorkPackages::EagerLoading::Base` to accept additional attributes and assign them to the local accessors if available. ➝ https://github.com/opf/openproject/pull/11783/commits/043e13a6dc222d74e49653b9bc141bf89eb79fc9 

This still uses `respond_to?` in this initializer. Please let me know if you had something different in mind.

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
```diff
@@ -0,0 +1,245 @@
+
+  ## Retrieve Work Packages for Different Timestamps
+
+  The work-packages API supports a `timestamps` parameter to gather information about a single work package or a collection work packages for several points in time.
```
Missing word:

> collection work packages

to

> collection of work packages

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
````diff
@@ -0,0 +1,245 @@
+  GET /api/v3/work_packages?timestamps=2022-01-01T00:00:00Z,PT0S
+
+  GET /api/v3/work_packages/123?timestamps=2022-01-01T00:00:00Z,PT0S
+  ```
````
Please turn these into two separate code blocks. Otherwise, they will be displayed in the same line.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
````diff
@@ -0,0 +1,245 @@
+  GET /api/v3/work_packages/123?timestamps=2022-01-01T00:00:00Z,PT0S
+  ```
+
+  Each timestamp should be given as [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) string, either as absolute date and time with timezone, e.g. `"2022-01-01T00:00:00Z"`, or as relative timestamp utilizing the [ISO-8601-Duration](https://en.wikipedia.org/wiki/ISO_8601#Durations) format, e.g. `"P-1Y"`, which is composed of an initial `"P"` for "Period", and a duration. `"P-1Y"` is interpreted a ...
````
Minor changes:

> Each timestamp should be given as [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) string, either as absolute date and time with timezone, e.g. `"2022-01-01T00:00:00Z"`, or as relative timestamp utilizing the

to

> ach timestamp should be given as an [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601) string, either an absolute date and time with timezone, e.g. `"2022-01-01T00:00:00Z"`, or a relative timestamp utilizing the

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
```diff
@@ -0,0 +1,245 @@
+
+  The timestamps should be given in ascending temporal order, such that the first timestamp is the **baseline** timestamp, and the last timestamp is the **current** timestamp.
+
+  The information of the work packages at the different `timestamps` is returned in the `"_embedded"` section `"attributesByTimestamp"`.
```
Instead of diving into the `attributesByTimestamp` here, I'd rather have an overview over the parts that are added compared to a non timestamp based work package collections: ' _meta' with the three properties and the `attributesByTimestamp` and then say that details can be found below in the various sections.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
```diff
@@ -0,0 +1,245 @@
+
+  The work-packages API supports filtering the query results by one or several search criteria. See: [Filters](../filters)
+
+  To find out whether a work package matches the given set of filter criteria, check `"matchesFilters"` in the `"_meta"` section for each timestamp:
```
> matches the given set of filter criteria

I'd emphasis that it is matching it at that timestamp.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
````diff
@@ -0,0 +1,245 @@
+  }
+  ```
+
+  In the above example, the work package matches the filter criteria at the timestamp `"PT0S"`, but does not match the filter criteria at the timestamp `"2022-01-01T00:00:00Z"`. The work package is included in the returned collection because it matches the filter criteria at least at one of the given timestamps.
````
I'd emphasis that it can also be the other way around so that the work package does not match at the timestamp of 'now' but at the other one.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
````diff
@@ -0,0 +1,245 @@
+  }
+  ```
+
+  In the above example, the work package exists at the timestamp `"PT0S"`, but has not existed at the timestamp `"2022-01-01T00:00:00Z"`. The work package is included in the returned collection because it has existed at least at one of the given timestamps.
````
I again don't think it waisted to explicitly point out that it could be the other way around. Although, that would require a weird kind of timestamps as we don't support soft deletion. So maybe only the lacking support of the soft deletion should be noted here.

fiedl replied:
➝ https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e

ulferts commented on docs/api/apiv3/tags/baseline_comparisons.yml:
```diff
@@ -0,0 +1,245 @@
+          "_meta": {
+            "matchesFilters": true,
+            "exists": true,
+            "timestamp": "PT0S"
```
The `timestamp` property is never explicitly explained. It should be rater self explanatory, but it can't hurt to have one sentence regarding it.

fiedl replied:
Previously I had not mentioned the `timestamps` property, because we weren't completely sure that we would keep the `timestamps` property. The information is redundant in the sense that the array order matches the order of the requested timestamps; so we already know the timestamp.

However, I like adding the `timestamp`  property explicitly to have a check and also for clarity.

I've added a short explanation for the `timestamp` property in https://github.com/opf/openproject/pull/11783/commits/5a33ab64d384e89c6da23412661af7e4aea2182e.

ulferts reviewed (changes requested):
Hey @fiedl ,

please find my remarks below. When reviewing, I skipped a couple of files simply because the will be altered by #11783 soon. But in total, the change is fine and only minor adjustments need to be made before it can be merged. Good job.

fiedl committed:
Timestamp: extracting redundant code into method

caught by Jens Ulferts' review

https://github.com/opf/openproject/pull/11783#discussion_r1072377773

fiedl committed:
baseline api: minor corrections after review

review by Jens Ulferts

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: refactoring nomenclature in collection representer

`query` refers to the `Query` object. `query_params` refers to a hash of query parameters.

https://github.com/opf/openproject/pull/11783#discussion_r1116731624

fiedl committed:
baseline api: move eager-loading-wrapper options assignment into base initializer

as suggested in review: https://github.com/opf/openproject/pull/11783#discussion_r1116783754

fiedl committed:
Merge 'dev' into https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: rubocop refactoring

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: fixing scenario in which only a historic timestamp is requested at which the work package does not exist

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: adding missing specs for wrapping multiple work packages

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: adding specs for caching

https://github.com/opf/openproject/pull/11783#discussion_r1116735903

fiedl committed:
baseline api: remove redundant `timestamp` attribute

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: fixing test issue with relative timestamps

https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: removing `timestamps` from `json_cache_key` for now.

The additional properties are already marked with `uncacheable: true`.

Re-enabling caching for certain scenarios is a follow-up step: https://github.com/fiedl/openproject/issues/4

https://github.com/opf/openproject/pull/11783#discussion_r1116735903

fiedl committed:
empty commit to re-trigger CI

fiedl committed:
baseline api: improving documentation

as requested in review by Jens Ulferts: https://github.com/opf/openproject/pull/11783#discussion_r1116795169

fiedl committed:
Merge 'dev' into https://github.com/opf/openproject/pull/11783

fiedl committed:
baseline api: adding missing specs for `Journable::WithHistoricAttributes`

https://github.com/opf/openproject/pull/11783

ulferts reviewed (approved):
@fiedl everything looking good now. Thanks for all the work you put in.
